### PR TITLE
REST API Basics

### PR DESCRIPTION
**Pull request by DmitryAkulich33**
added: dao (not all), domain , mapper, sql scripts

### CODE REVIEW

**Review by sznv — CHANGES_REQUESTED** [review]

File: rest/build/resources/main/sql/3_create_tables.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
+USE gift_certificate;
+
+create table tag
+(
+    id_tag INTEGER NOT NULL AUTO_INCREMENT,
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/build/resources/main/sql/3_create_tables.sql:5]: Done!

File: rest/src/main/java/com/epam/esm/Application.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.epam.esm;
+
+public class Application {
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/Application.java:3]: Done!

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+public class CertificateController {
+    private final CertificateService certificateService;
+
+    @Autowired
+    public CertificateController(CertificateService certificateService) {
+        this.certificateService = certificateService;
+    }
+
+    @DeleteMapping(path ="/{id}")
+    public List<Certificate> deleteById (@PathVariable int id){
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java:23]: Done!

File: rest/src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+public class CertificateDaoImpl implements CertificateDao {
+    private final JdbcTemplate template;
+    private final RowMapper<Certificate> rowMapper;
+
+    private static final String ADD_CERTIFICATE = "INSERT INTO certificate " +
+            "(name_certificate, description, price, creation_date, duration) VALUES(?,?,?,?,?)";
+    private static final String EDIT_BY_ID = "UPDATE certificate SET name_certificate=?, description=?, price=?, " +
+            "update_date=?, duration=? WHERE id_certificate=? AND lock_certificate=0";
+    private static final String LOCK_BY_ID = "UPDATE certificate SET lock_certificate=1 WHERE id_certificate=?";
+    private static final String FIND_BY_ID = "SELECT * FROM certificate WHERE id_certificate=? AND lock_certificate=0";
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java:23]: Done!

File: rest/src/main/java/com/epam/esm/config/SpringConfig.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.web.servlet.view.InternalResourceViewResolver;
+import org.springframework.web.servlet.view.JstlView;
+
+import javax.sql.DataSource;
+
+@Configuration
+@EnableWebMvc
+@PropertySource("classpath:jdbc.properties")
+@ComponentScan(basePackages = "com.epam.esm")
+public class SpringConfig implements WebMvcConfigurer {
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/config/SpringConfig.java]: Done

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import com.epam.esm.service.CertificateService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping(value = "/certificate", produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java]: Done

[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java]: Done

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java]: Done

File: rest/src/main/java/com/epam/esm/config/SpringConfig.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        ds.setJdbcUrl(environment.getRequiredProperty("jdbc.url"));
+        ds.setUsername(environment.getRequiredProperty("jdbc.username"));
+        ds.setPassword(environment.getRequiredProperty("jdbc.password"));
+        return ds;
+    }
+
+    @Bean
+    public JdbcTemplate jdbcTemplate() {
+        return new JdbcTemplate(dataSource());
+    }
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/config/SpringConfig.java]: Done

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+public class CertificateController {
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java:16]: Done

[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java:16]: Done

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    @DeleteMapping(path ="/{id}")
+    public List<Certificate> deleteById (@PathVariable int id){
+        certificateService.lockById(id);
+        return certificateService.findAll();
+    }
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public List<Certificate> add (@RequestBody Certificate certificate){
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java]: Done

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java:16]: Done

File: rest/src/main/java/com/epam/esm/controller/CertificateController.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        return certificateService.findAll();
+    }
+
+
+    @GetMapping("/{id}")
+    public Certificate findById(@PathVariable int id) {
+        return certificateService.findById(id);
+    }
+
+    @GetMapping(path = "/certificates")
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/CertificateController.java]: Done

File: rest/src/main/java/com/epam/esm/controller/TagController.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
+package com.epam.esm.controller;
+
+import com.epam.esm.domain.Certificate;
```
[the remark this answers is not in the document]
Comment (DmitryAkulich33) [reply rest/src/main/java/com/epam/esm/controller/TagController.java]: Done

**Review by sznv — APPROVED** [review]

**Pull request merged**